hal/src/rtl872x/pwm_hal.cpp

```diff
@@ -373,10 +371,10 @@ void hal_pwm_set_resolution(uint16_t pin, uint8_t resolution) {
         }
     }
 
-    uint32_t prePeriodCycles = (1 << pwmInfo[instance].resolution[channel]);
+    uint32_t prePeriodCycles = (1 << pwmInfo[instance].channels[channel].resolution);
     uint32_t newPeriodCycles = (1 << resolution);
-    pwmInfo[instance].resolution[channel] = resolution;
-    pwmInfo[instance].values[channel] = ((float)pwmInfo[instance].values[channel] / prePeriodCycles) * newPeriodCycles;
+    pwmInfo[instance].channels[channel].resolution = resolution;
+    pwmInfo[instance].channels[channel].value = ((float)pwmInfo[instance].channels[channel].value / prePeriodCycles) * newPeriodCycles;

```

Comment:
When a new resolution is set, should we recalculate the ARR and CCMRX register values (ie duty cycle) to update the new values?

Reply:
If a pin is already running a non-zero PWM output, and the resolution is changed... we don't have docs that say what is expected to happen.  I don't think the duty cycle value should be re-adjusted though (that's up to the user), but the frequency probably should remain the same (if it already doesn't).  We probably should have a note that says something like, "It's recommended to set the resolution before calculating the duty cycle value and setting the PWM to a non-zero value." ?

Reply:
We will store the new resolution and duty cycle value in the PWM HAL, but it wont be applied to the timer registers and the PWM output duty cycle / frequency wont change. 
The user will have to call `analogWrite()` to get their duty cycle/frequency changes to take effect. So there is no new behavior here, but yeah could be documented better?

Reply:
> We will store the new resolution and duty cycle value in the PWM HAL, but it wont be applied to the timer registers and the PWM output duty cycle / frequency wont change.
> The user will have to call analogWrite() to get their duty cycle/frequency changes to take effect. So there is no new behavior here, but yeah could be documented better?

That's the expected behavior.